homeassistant/components/solaredge/sensor.py

```diff
@@ -79,30 +93,67 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
         _LOGGER.error("Could not retrieve details from SolarEdge API")
         return
 
-    # Create solaredge data service which will retrieve and update the data.
-    data = SolarEdgeData(hass, api, site_id)
+    # Create sensor factory that will create sensors based on sensor_key.
+    sensor_factory = SolarEdgeSensorFactory(platform_name, site_id, api)
 
     # Create a new sensor for each sensor type.
     entities = []
     for sensor_key in config[CONF_MONITORED_CONDITIONS]:
-        sensor = SolarEdgeSensor(platform_name, sensor_key, data)
+        sensor = sensor_factory.create_sensor(sensor_key)
         entities.append(sensor)
 
     add_entities(entities, True)
 
 
+class SolarEdgeSensorFactory:
+    """Factory which creates sensors based on the sensor_key."""
+
+    def __init__(self, platform_name, site_id, api):
+        """Initialize the factory."""
+        self.platform_name = platform_name
+
+        self.overview_data_service = SolarEdgeOverviewDataService(api, site_id)
+        self.details_data_service = SolarEdgeDetailsDataService(api, site_id)
+        self.inventory_data_service = SolarEdgeInventoryDataService(api,
+                                                                    site_id)
+        self.power_flow_data_service = SolarEdgePowerFlowDataService(api,
+                                                                     site_id)
+
+    def create_sensor(self, sensor_key):
+        """Create and return a sensor based on the sensor_key."""
+        sensor = None
+
+        if sensor_key == 'site_details':
+            sensor = SolarEdgeDetailsSensor(self.platform_name, sensor_key,
+                                            self.details_data_service)
+        elif sensor_key in ['meters', 'sensors', 'gateways',
+                            'batteries', 'inverters']:
+            sensor = SolarEdgeInventorySensor(self.platform_name, sensor_key,
+                                              self.inventory_data_service)
+        elif sensor_key in ['power_consumption', 'solar_power', 'grid_power',
+                            'storage_power']:
+            sensor = SolarEdgePowerFlowSensor(self.platform_name, sensor_key,
+                                              self.power_flow_data_service)
+        else:
+            sensor = SolarEdgeOverviewSensor(self.platform_name, sensor_key,
+                                             self.overview_data_service)
+        return sensor
+
+
 class SolarEdgeSensor(Entity):
-    """Representation of an SolarEdge Monitoring API sensor."""
+    """Abstract class for a solaredge sensor."""
 
-    def __init__(self, platform_name, sensor_key, data):
+    def __init__(self, platform_name, sensor_key, data_service):
         """Initialize the sensor."""
+        self.entity_id = '{}.{}'.format(platform_name, sensor_key).lower()
```

Comment:
I wanted to leave the option open to change the name of the sensors later but leave the entity_id as it is to prevent breaking changes in configuration files. Might not be an issue as much and have removed it.